sdk/eventhubs/azure-messaging-eventhubs/src/consumer_client.cpp

```diff
@@ -48,31 +48,73 @@ namespace Azure { namespace Messaging { namespace EventHubs {
+    }
+    while (!m_connections.empty())
+    {
+      m_connections.erase(m_connections.begin());
+    };
+  }
+
+  Azure::Core::Amqp::_internal::Connection ConsumerClient::CreateConnection(
+      std::string const& partitionId)
+  {
+    ConnectionOptions connectOptions;
+    connectOptions.ContainerId
+        = "Consumer for " + m_consumerClientOptions.ApplicationID + " on " + partitionId;
+    connectOptions.EnableTrace = _detail::EnableAmqpTrace;
+    connectOptions.AuthenticationScopes = {"https://eventhubs.azure.net/.default"};
+
+    // Set the user agent related properties in the connectOptions based on the package
+    // information and application ID.
+    _detail::EventHubsUtilities::SetUserAgent(
+        connectOptions, m_consumerClientOptions.ApplicationID);
+
+    return Azure::Core::Amqp::_internal::Connection{
+        m_fullyQualifiedNamespace, m_credential, connectOptions};
+  }
+
+  void ConsumerClient::EnsureConnection(std::string const& partitionId)
+  {
+    std::unique_lock<std::mutex> lock(m_sessionsLock);
+    if (m_connections.find(partitionId) == m_connections.end())
+    {
+      m_connections.emplace(partitionId, CreateConnection(partitionId));
+    }
+  }
+
+  Azure::Core::Amqp::_internal::Session ConsumerClient::CreateSession(
+      std::string const& partitionId)
+  {
+    SessionOptions sessionOptions;
+    sessionOptions.InitialIncomingWindowSize
+        = static_cast<uint32_t>(std::numeric_limits<int32_t>::max());
+
+    return m_connections.at(partitionId).CreateSession(sessionOptions);
+  }
+
   void ConsumerClient::EnsureSession(std::string const& partitionId)
   {
+    EnsureConnection(partitionId);
+    std::unique_lock<std::mutex> lock(m_sessionsLock);
     if (m_sessions.find(partitionId) == m_sessions.end())
     {
-      ConnectionOptions connectOptions;
-      connectOptions.ContainerId = m_consumerClientOptions.ApplicationID;
-      connectOptions.EnableTrace = true;
-      connectOptions.AuthenticationScopes = {"https://eventhubs.azure.net/.default"};
-
-      // Set the user agent related properties in the connectOptions based on the package
-      // information and application ID.
-      _detail::EventHubsUtilities::SetUserAgent(
-          connectOptions, m_consumerClientOptions.ApplicationID);
-
-      Connection connection(m_fullyQualifiedNamespace, m_credential, connectOptions);
-      SessionOptions sessionOptions;
-      sessionOptions.InitialIncomingWindowSize
-          = static_cast<uint32_t>(std::numeric_limits<int32_t>::max());
-
-      Session session{connection.CreateSession(sessionOptions)};
-      m_sessions.emplace(partitionId, session);
+      m_sessions.emplace(partitionId, CreateSession(partitionId));
     }
   }
 
-  Azure::Core::Amqp::_internal::Session ConsumerClient::GetSession(std::string const& partitionId)
+  Azure::Core::Amqp::_internal::Session ConsumerClient::GetSession(
+      std::string const& partitionId = {})
   {
     return m_sessions.at(partitionId);
```

Comment:
This also does access `m_sessions` without acquiring `m_sessionsLock`

Reply:
Ah, if this is private, then it could be ok, depending on the places where it gets called from.

Reply:
If so, then maybe if it was named something like "GetSessionWithoutAcquiringMutex", or something else along these lines, that emphasizes that all you see is intentional, a code readers like me would not have this reaction. Or maybe it is ok to leave the old name, I am leaving this up to you.

Reply:
Or maybe a comment just above that line, that the method is private and the mutex is taken care of would do it.